File: src/screens/__tests__/recordLabels.test.ts
import { labelsFor } from '../meeting/recordLabels';

/**
 * What a typed item wears. The defects it guards: a free-tier item (no record) growing labels
 * it never earned, "Open" shown as if it were news, and a day label the phone and the screen
 * disagree about.
 */
describe('labelsFor', () => {
  it('shows nothing for an unclassified item', () => {
    expect(labelsFor(null)).toEqual({});
    expect(labelsFor({ itemType: null, status: null, dateNorm: null })).toEqual({});
  });

  it('shows the type, a status only when a later turn changed it, and the pinned day', () => {
    expect(labelsFor({ itemType: 'request', status: 'contradicted', dateNorm: null })).toEqual({
      type: 'Request',
      status: 'Contradicted',
    });
    expect(labelsFor({ itemType: 'commitment', status: 'open', dateNorm: Date.UTC(2026, 8, 18) })).toEqual({
      type: 'Commitment',
      day: 'Fri 18 Sep',
    });
  });

  it('says so when the model was not sure', () => {
    expect(labelsFor({ itemType: 'uncertain', status: 'open', dateNorm: null })).toEqual({ type: 'Not sure' });
  });
});

File: android/app/src/test/java/com/innocorelabs/verbale/pipeline/RecordLabelsTest.kt
package com.innocorelabs.verbale.pipeline

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * What a typed item wears in an export. The defects it guards: a free-tier row growing labels it
 * never earned, "Open" printed as if it were news, a day label the document and the screen
 * disagree about, and "Sept". Shared with recordLabels.test.ts row for row.
 */
class RecordLabelsTest {
  private val golden: JSONObject by lazy {
    val f = listOf("../../cpp/tests/golden/record_labels.json", "../cpp/tests/golden/record_labels.json", "cpp/tests/golden/record_labels.json")
      .map { File(it) }.firstOrNull { it.exists() }
      ?: error("record_labels.json not found from ${File(".").absolutePath}")
    JSONObject(f.readText())
  }

  private fun JSONObject.str(k: String): String? = if (isNull(k)) null else getString(k)

  @Test fun everyRowOfTheGoldenTable() {
    val cases = golden.getJSONArray("cases")
    assertTrue(cases.length() > 7)
    for (i in 0 until cases.length()) {
      val c = cases.getJSONObject(i)
      val dateNorm = if (c.isNull("dateNorm")) null else c.getLong("dateNorm")
      val l = RecordLabels.labelsFor(c.str("itemType"), c.str("status"), dateNorm)
      assertEquals(c.getString("name"), c.str("type"), l.type)
      assertEquals(c.getString("name"), c.str("statusLabel"), l.status)
      assertEquals(c.getString("name"), c.str("day"), l.day)
      assertEquals(c.getString("name"), c.getString("suffix"), RecordLabels.suffix(c.str("itemType"), c.str("status"), dateNorm))
    }
  }
}
